### PR TITLE
Add user principal 464

### PR DESCRIPTION
**Pull request by Incarnation-p-lee**
* Add user principal to filter, #464. 

Signed-off-by: Pan Li <panli@microsoft.com>

### CODE REVIEW

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java [hunk 1]
```diff
@@ -25,42 +32,99 @@
...
      * @param properties of ${@link AADB2CProperties}.
      * @return the instance of ${@link AADB2CJWTProcessor}.
-     * @throws AADB2CAuthenticationException when failed to create ${@link AADB2CJWTProcessor} instance.
      */
-    private AADB2CJWTProcessor getAADB2CJwtParser(@URL String url, @NonNull AADB2CProperties properties)
-            throws AADB2CAuthenticationException {
-        urlToJwtParser.putIfAbsent(url, new AADB2CJWTProcessor(url, properties));
+    private AADB2CJWTProcessor getAADB2CJwtProcessor(@URL String url, @NonNull AADB2CProperties properties) {
+        // TODO(pan): url should exclude state and nonce UUID part.
+        URL_TO_JWT_PARSER.putIfAbsent(url, new AADB2CJWTProcessor(url, properties));
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: `computeIfAbsent`? We should delay creating new instance.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java [hunk 2]
```diff
@@ -25,42 +32,99 @@
...
-        super.validateAuthentication(request);
+    /**
+     * Validate the reply state from AAD B2C, the state compose of tow parts with format 'UUID-RequestURL',
+     * for example: 461e6d45-37cf-4a8f-9fd8-086b98c8abfb-http://localhost:8080/
+     *
+     * @param state encoded in policy URL and replied by AAD B2C.
+     * @return the request URL.
+     */
+    private String validateState(String state) throws AADB2CAuthenticationException {
+        final int uuidLength = UUID.randomUUID().toString().length();
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: Do we need to do it every time?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java [hunk 3]
```diff
@@ -25,42 +32,99 @@
...
-        return urlToJwtParser.get(url);
+        return URL_TO_JWT_PARSER.get(url);
     }
 
-    @Override
-    public void handle(HttpServletRequest request, HttpServletResponse response, AADB2CProperties properties)
-            throws AADB2CAuthenticationException {
-        super.validateAuthentication(request);
+    /**
+     * Validate the reply state from AAD B2C, the state compose of tow parts with format 'UUID-RequestURL',
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: What's the story behind it? Is there any official documentation link we can put here?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CURL.java [hunk 4]
```diff
@@ -14,11 +15,25 @@
...
+     * The ${@link Set} of generated ${@link UUID} for state field.
+     */
+    @Getter
+    private static Set<String> stateIDSet = new ConcurrentSkipListSet<>();
+
+    /**
+     * The ${@link Set} of generated ${@link UUID} for nonce field.
+     */
+    @Getter
+    private static Set<String> nonceIdSet = new ConcurrentSkipListSet<>();
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CURL.java]: I see memory leak here.
Is there a built-in solution to generate nonce?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        this.audiences = jwtClaimsSet.getAudience();
+        this.expireDate = jwtClaimsSet.getExpirationTime();
+        this.notValidBeforeDate = jwtClaimsSet.getNotBeforeTime();
+        this.issueAtDate = jwtClaimsSet.getIssueTime();
+        this.claim = jwtClaimsSet.getClaims();
+        this.code = code;
+
+        this.userObjectId = jwtClaimsSet.getSubject(); // Subject populates the object ID of the user in the directory.
+
+        this.surname = (String) getClaim().get(ATTRIBUTE_SURNAME);
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java]: Cache the result of `getClaim()`?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java (hunk 1 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: yes, should be computeIfAbsent.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java (hunk 2 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: The document recommend to use UUID or something like page user stay. Consider to remove UUID as the nonce can cover this part.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CURL.java (hunk 4 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CURL.java]: It looks not easy to handle nonce by ourselves, will investigate some existed solution about this.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java (hunk 5 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java]: seems unnecessary here, as just a Getter.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CEntryPoint.java [hunk 6]
```diff
@@ -28,9 +29,17 @@ public AADB2CEntryPoint(@NonNull AADB2CProperties properties) {
...
-            IOException {
-        final String requestURL = request.getRequestURL().toString();
+    public void commence(HttpServletRequest request, HttpServletResponse response, AuthenticationException e)
+            throws IOException {
+        String requestURL = request.getRequestURL().toString();
+        final URL url = new URL(requestURL);
+
+        // By default, request URL will be '/error' when exception handling, here handle this endpoint to
+        // policy redirect URL instead of redirect to '/error' directly.
+        if (url.getPath().equalsIgnoreCase("/error")) {
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CEntryPoint.java]: I am okay with POC here. But for the final solution, I think we need more discussion.

Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CEntryPoint.java]: Yes, absolutely.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenarioHandler.java [hunk 7]
```diff
@@ -20,7 +21,7 @@
  * spring security context if success, or take care the error response and throw exception.
  * <p>
  * ${@link AADB2CFilterLogoutSuccessHandler} will obtain the result from AAD B2C, it will do nothing if success,
- * or warn the user that the token failed to expire.
+ * or warn the user that the token failed to expireDate.
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterScenarioHandler.java]: Typo?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CLogoutSuccessHandler.java [hunk 8]
```diff
@@ -54,7 +54,7 @@ public void onLogoutSuccess(HttpServletRequest request, HttpServletResponse resp
         final String redirectURL = AADB2CURL.getOpenIdLogoutURL(b2cProperties, getLogoutSuccessURL(), request);
 
         log.debug("Redirect to AAD B2C URL {} for expiring token.", redirectURL);
-        log.debug("Expire AAD B2C token success will redirect to URL {}.", getLogoutSuccessURL());
+        log.debug("Expire AAD B2C token successful, will redirect to URL {}.", getLogoutSuccessURL());
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CLogoutSuccessHandler.java]: -> `AAD B2C token has been successfully expired. Redirecting to URL {}` ?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+import java.util.Map;
+
+@Getter
+@ToString
+public class UserPrincipal implements Serializable {
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    private static final long serialVersionUID = 2194984731878145577L;
+
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java]: Do we need blank lines between all these constants?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    private static final String ATTRIBUTE_CODE_HASH = "c_hash";
+
+    private static final String ATTRIBUTE_NONCE = "nonce";
+
+    private static final String ATTRIBUTE_ID_PROVIDER = "idp";
+
+    // JWT token
+    private final String keyId;
+
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java]: Similarly, do we need blank lines between these fields?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java [hunk 11]
```diff
@@ -25,42 +31,94 @@
...
+    public void handle(HttpServletRequest request, HttpServletResponse response, AADB2CProperties properties)
+            throws AADB2CAuthenticationException, IOException {
+        Assert.notNull(properties, "AADB2CProperties should not be null.");
+
+        final Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+
+        if (isAuthenticated(auth)) {
+            return;
+        } else if (auth instanceof PreAuthenticatedAuthenticationToken) {
+            log.debug("User principal {} not authenticated.", ((UserPrincipal) auth.getPrincipal()).getDisplayName());
```
Comment (xscript) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: -> `User {} is not authenticated.`?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java (hunk 9 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java]: Follow the convention of this repo, for readability.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        this.nonce = (String) getClaim().get(ATTRIBUTE_NONCE);
+        this.identityProvider = (String) getClaim().get(ATTRIBUTE_ID_PROVIDER);
+
+        this.legalAgeGroup = "";
+
+        List<String> emails;
+        Boolean isNew;
+
+        try {
+            isNew = (Boolean) getClaim().get(ATTRIBUTE_NEW_USER);
```
[the remark this answers is not in the document]
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java]: The attribution type is `Boolean`, which is not suitable for `parseBoolean`.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+        } catch (ClassCastException ignore) {
+            isNew = false;
+        }
+
+        try {
+            emails = (List<String>) OBJECT_MAPPER.readValue(getClaim().get(ATTRIBUTE_EMAILS).toString(), List.class);
```
[the remark this answers is not in the document]
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java]: the same as below, List<String> there.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java (hunk 11 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: Updated log message.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java [hunk 14]
```diff
@@ -25,42 +31,94 @@
 public class AADB2CFilterSignUpOrInHandler extends AbstractAADB2CFilterScenarioHandler
         implements AADB2CFilterScenarioHandler {
 
+    private final RedirectStrategy redirectStrategy = new DefaultRedirectStrategy();
+
     /**
      * Mapping configuration URL to ${@link AADB2CJWTProcessor}.
      */
-    private final ConcurrentMap<String, AADB2CJWTProcessor> urlToJwtParser = new ConcurrentHashMap<>();
+    private static final ConcurrentMap<String, AADB2CJWTProcessor> URL_TO_JWT_PARSER = new ConcurrentHashMap<>();
```
Comment (warrenzhu25) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: This is not a constant. Original naming is fine.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java [hunk 15]
```diff
@@ -25,42 +31,94 @@
...
+//
+//        final String expectedNonce = request.getSession().getAttribute(AADB2CURL.ATTRIBUTE_NONCE).toString();
+//
+//        if (!nonce.equals(expectedNonce)) {
+//            throw new AADB2CAuthenticationException("The claim nonce has unexpected content: " + nonce);
+//        }
+    }
 
-        Assert.notNull(properties, "AADB2CProperties should not be null.");
+    private boolean isAuthenticated(Authentication auth) {
```
Comment (warrenzhu25) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: could be static

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+        this.notValidBeforeDate = jwtClaimsSet.getNotBeforeTime();
+        this.issueAtDate = jwtClaimsSet.getIssueTime();
+        this.claim = jwtClaimsSet.getClaims();
+        this.code = code;
+
+        this.userObjectId = jwtClaimsSet.getSubject(); // Subject populates the object ID of the user in the directory.
+
+        this.surname = (String) getClaim().get(ATTRIBUTE_SURNAME);
+        this.street = (String) getClaim().get(ATTRIBUTE_STREET_ADDRESS);
+        this.stateOrProvince = (String) getClaim().get(ATTRIBUTE_STATE);
```
Comment (warrenzhu25) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java]: This class has too many fields. 
1. Is it possible to split this?
2. Could we use map to store all attributes?

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java (hunk 15 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: will move to abstract class, as other implementation may leverage this.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java (hunk 14 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CFilterSignUpOrInHandler.java]: make sense, will revert this.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java (hunk 16 above)
Comment (Incarnation-p-lee) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java]: Mostly are from B2C pre-defined attribution, split may involve other class which may unnecessary here, and the `getClaim()` holds the map respectively.

**Pull request merged**